### PR TITLE
CRM-19300 : allow external id for custom fields import

### PR DESCRIPTION
**Pull request by liedekef**
- [CRM-19300: Allow external id to be used for custom multi-valued fields import](https://issues.civicrm.org/jira/browse/CRM-19300)

### CODE REVIEW

File: CRM/Custom/Import/Form/MapField.php [hunk 1]
```diff
@@ -6,7 +6,7 @@
 class CRM_Custom_Import_Form_MapField extends CRM_Contact_Import_Form_MapField {
   protected $_parser = 'CRM_Custom_Import_Parser_Api';
   protected $_mappingType = 'Import Multi value custom data';
-  protected $_highlightedFields = array();
+//  protected $_highlightedFields = array();
```
Comment (seamuslee001) [inline CRM/Custom/Import/Form/MapField.php]: Not sure why this is commented out 

I would have thought later down you could have done $this->_highlightedfields[] = ... and left the page assignment to $this->_highlightedfields.

Comment (liedekef) [inline CRM/Custom/Import/Form/MapField.php]: I just looked how it was done in Contact import and tried to stay in line with that code.

Comment (liedekef) [inline CRM/Custom/Import/Form/MapField.php]: If you like to see this differently, I can change it of course. I'll fix the styling issues already

**Review by colemanw — APPROVED** [review]
Code looks good to me; haven't tested the functionality.

File: CRM/Custom/Import/Form/MapField.php [hunk 2]
```diff
@@ -24,6 +23,9 @@ public function preProcess() {
     $this->_columnCount = $this->get('columnCount');
     $this->assign('columnCount', $this->_columnCount);
     $this->_dataValues = $this->get('dataValues');
+    $highlightedFields = array();
+    $highlightedFields[] = 'contact_id';
+    $highlightedFields[] = 'external_identifier';
```
Comment (monishdeb) [inline CRM/Custom/Import/Form/MapField.php]: @liedekef can be optimise into 
```
 $highlightedFields = array('contact_id', 'external_identifier');
```

File: CRM/Custom/Import/Form/MapField.php [hunk 3]
```diff
@@ -72,16 +74,13 @@ public static function formRule($fields) {
...
-            $errors['_qf_default'] = '';
-          }
-          $errors['_qf_default'] .= ts('Missing required field: %1', array(1 => $title));
+
+      // check either contact id or external identifier
+      if (!in_array('contact_id', $importKeys) && !in_array('external_identifier', $importKeys)) {
+        if (!isset($errors['_qf_default'])) {
+          $errors['_qf_default'] = '';
         }
+        $errors['_qf_default'] .= ts('Missing required field: %1', array(1 => ts('Contact ID')));
```
Comment (monishdeb) [inline CRM/Custom/Import/Form/MapField.php]: Should we change it to 
```
ts('Missing required field: %1', array(1 => ts('Contact ID or External Identifier')));
```
 ?

**Review by monishdeb — CHANGES_REQUESTED** [review]

**Comment by monishdeb** [comment]
Although it is not related to the changes, but found another strict warning 
image

Will submit a PR for it later

File: CRM/Custom/Import/Form/MapField.php (hunk 2 above)
Comment (liedekef) [inline CRM/Custom/Import/Form/MapField.php]: I agree, but again: I just the same form from the contact import code. Do you want me to change it here already?

File: CRM/Custom/Import/Form/MapField.php (hunk 3 above)
Comment (liedekef) [inline CRM/Custom/Import/Form/MapField.php]: It's the same remark I made above :-) But indeed, I agree. Should I change it?

**Comment by liedekef** [comment]
For that strict-warning, I opened a ticket in the past but got no reaction to it.

**Comment by monishdeb** [comment]
@liedekef when you override a parent class function (as in here ```CRM_Custom_Import_Parser::run(...)``` ) we need to ensure that the function parameters are same in number and type as of parent function parameters (as in here ```CRM_Contact_Import_Parser::run(...)``` ). And on function parameters mismatch it throws such kind of notice, however not fatal in nature. Here the fix that we need to ensure the parameter match  and ofcourse bring changes in the function definition too. Would you now like to submit a patch for it ? and what is the ticket number ?

File: CRM/Custom/Import/Form/MapField.php (hunk 3 above)
Comment (monishdeb) [inline CRM/Custom/Import/Form/MapField.php]: ya sure go ahead

File: CRM/Custom/Import/Form/MapField.php (hunk 2 above)
Comment (monishdeb) [inline CRM/Custom/Import/Form/MapField.php]: Yes please :)

File: CRM/Custom/Import/Form/MapField.php (hunk 3 above)
Comment (liedekef) [inline CRM/Custom/Import/Form/MapField.php]: I changed "upstream" by "customfields", but for the rest I followed your steps and it should be ok now :-)

**Comment by liedekef** [comment]
I new the strict-warning was related to the number of parameters, but not really a bug. But currently I don't feel confident to correct that. But the ticket number is: CRM-19281

**Review by monishdeb — APPROVED** [review]

**Comment by monishdeb** [comment]
@liedekef thanks for your additional fix. Tested the patch, works like a charm :)

**Pull request merged**